src/common/utilities/FormulaEvaluation.ts

```diff
@@ -0,0 +1,641 @@
+                let arrayString = token.value as string;
+                // Remove leading and trailing square brackets
+                arrayString = arrayString.slice(1, -1);
+                // Split the string by commas and trim whitespace
+                const arrayElements = arrayString.split(',').map(element => element.trim());
+                stack.push(new ArrayLiteralNode(arrayElements));
+            } else if (token.type === "FUNCTION") {
+                const arity = this._getFnArity(token.value as string);
+                const operands: ASTNode[] = [];
+                while (operands.length < arity) {
+                    operands.push(stack.pop() as ASTNode);
+                }
+                stack.push({ type: token.type, value: token.value, operands: operands.reverse() });
+            }
+        }
+
+        // At this stage, the stack should contain a single node representing the root of the AST
+        return stack[0] as ASTNode;
+    }
+
+    public evaluateASTNode(node: ASTNode | ArrayLiteralNode | string | number, context: { [key: string]: any } = {}): any {
+
+        if (!node) return 0;
+
+        if (typeof node === "object" && !(Object.prototype.hasOwnProperty.call(node, 'type') && Object.prototype.hasOwnProperty.call(node, 'value'))) {
+            return node;
+        }
+
+        // Each element in an array literal is evaluated recursively
+        if (node instanceof ArrayLiteralNode) {
+            const evaluatedElements = (node as ArrayLiteralNode).elements.map(element => this.evaluateASTNode(element, context));
+            return evaluatedElements;
+        }
+
+        // Number and string literals are returned as-is
+        if (typeof node === "number" || typeof node === "string") {
+            return node;
+        }
+
+        // Nodes with a type of NUMBER are parsed to a number
+        if (node.type === "NUMBER") {
+            const numVal = Number(node.value);
+            if (isNaN(numVal)) {
+                throw new Error(`Invalid number: ${node.value}`);
+            }
+            return numVal;
+        }
+
+        // Nodes with a type of BOOLEAN are parsed to a boolean
+        if (node.type === "BOOLEAN") {
+            return node.value === "true" ? 1 : 0;
+        }
+
+        // WORD and STRING nodes are returned as-is
+        if (node.type === "WORD" || node.type === "STRING") {
+            return node.value?.toString();
+        }
+
+        // VARIABLE nodes are looked up in the context object and returned
+        if (node.type === "VARIABLE") {
+            return context[(node.value as string).replace(/^[[@]?\$?([a-zA-Z_][a-zA-Z_0-9.]*)\]?/, '$1')] ?? null;
+        }
+
+        // OPERATOR nodes have their OPERANDS evaluated recursively, with the operator applied to the results
+        if (node.type === "OPERATOR" && ["+", "-", "*", "/", "==", "!=", "<>", ">", "<", ">=", "<=", "&&", "||", "%", "&", "|"].includes(node.value as string) && node.operands) {
```

Comment:
same as above: I'd suggest having constants.
Also - shouldn't a set of operators here and above be the same?

Reply:
This set of operators was slightly different from the preceding one as I convert ternary operator nodes (`?` and `:`) to a single `ternary` node with three operands in an earlier phase of the algorithm - therefore I know they won't occur at this point in the code. However, that also means it doesn't hurt to have them as part of the array. So as suggested, `operatorTypes` is now a const outside of the class 👍